File: backend/app/services/cells.py
"""Parse fenced ```python blocks out of LLM responses for notebook cells."""

import re

_FENCE = re.compile(r"```(?:python|py)\s*\n(.*?)```", re.DOTALL)


def extract_cells(markdown: str) -> list[dict[str, str]]:
    cells: list[dict[str, str]] = []
    for idx, match in enumerate(_FENCE.finditer(markdown)):
        code = match.group(1).rstrip()
        if code:
            cells.append({"index": str(idx), "language": "python", "source": code})
    return cells

File: frontend/src/lib/cells.ts
export type ParsedCell = { index: number; language: string; source: string };

const FENCE = /```(?:python|py)\s*\n([\s\S]*?)```/g;

export function parseCells(markdown: string): ParsedCell[] {
  const cells: ParsedCell[] = [];
  let match: RegExpExecArray | null;
  let idx = 0;
  FENCE.lastIndex = 0;
  while ((match = FENCE.exec(markdown)) !== null) {
    const source = match[1].replace(/\s+$/g, "");
    if (source) cells.push({ index: idx++, language: "python", source });
  }
  return cells;
}

export function splitProseAndCode(markdown: string): Array<
  | { kind: "prose"; text: string }
  | { kind: "code"; source: string }
> {
  const out: Array<{ kind: "prose"; text: string } | { kind: "code"; source: string }> = [];
  const re = /```(?:python|py)\s*\n([\s\S]*?)```/g;
  let cursor = 0;
  let match: RegExpExecArray | null;
  while ((match = re.exec(markdown)) !== null) {
    if (match.index > cursor) {
      const text = markdown.slice(cursor, match.index);
      if (text.trim()) out.push({ kind: "prose", text });
    }
    out.push({ kind: "code", source: match[1].replace(/\s+$/g, "") });
    cursor = re.lastIndex;
  }
  if (cursor < markdown.length) {
    const tail = markdown.slice(cursor);
    if (tail.trim()) out.push({ kind: "prose", text: tail });
  }
  return out;
}
